include/exadg/acoustic_conservation_equations/spatial_discretization/operators/grad_div_operator.h

```diff
@@ -0,0 +1,537 @@
+      integrator_flags_u.cell_integrate = dealii::EvaluationFlags::values;
+      integrator_flags_u.face_evaluate  = dealii::EvaluationFlags::values;
+      integrator_flags_u.face_integrate = dealii::EvaluationFlags::values;
+    }
+    else if(data_in.formulation == FormulationGradientDivergenceTerms::SkewSymmetric)
+    {
+      integrator_flags_p.cell_evaluate  = dealii::EvaluationFlags::gradients;
+      integrator_flags_p.cell_integrate = dealii::EvaluationFlags::gradients;
+      integrator_flags_p.face_evaluate  = dealii::EvaluationFlags::values;
+      integrator_flags_p.face_integrate = dealii::EvaluationFlags::values;
+
+      integrator_flags_u.cell_evaluate  = dealii::EvaluationFlags::values;
+      integrator_flags_u.cell_integrate = dealii::EvaluationFlags::values;
+      integrator_flags_u.face_evaluate  = dealii::EvaluationFlags::values;
+      integrator_flags_u.face_integrate = dealii::EvaluationFlags::values;
+    }
+    else
+    {
+      AssertThrow(false, dealii::ExcMessage("Not implemented."));
+    }
+  }
+
+
+  void
+  evaluate(BlockVectorType & dst, BlockVectorType const & src, double const time) const
+  {
+    do_evaluate(dst, src, true, time);
+  }
+
+  void
+  evaluate_add(BlockVectorType & dst, BlockVectorType const & src, double const time) const
+  {
+    do_evaluate(dst, src, false, time);
+  }
+
+private:
+  void
+  do_evaluate(BlockVectorType &       dst,
+              BlockVectorType const & src,
+              double const            time,
+              bool const              zero_dst_vector) const
+  {
+    if(data.formulation == FormulationGradientDivergenceTerms::Strong)
+    {
+      do_evaluate_formulation<FormulationGradientDivergenceTerms::Strong>(dst,
+                                                                          src,
+                                                                          time,
+                                                                          zero_dst_vector);
+    }
+    else if(data.formulation == FormulationGradientDivergenceTerms::Weak)
+    {
+      do_evaluate_formulation<FormulationGradientDivergenceTerms::Weak>(dst,
+                                                                        src,
+                                                                        time,
+                                                                        zero_dst_vector);
+    }
+    else if(data.formulation == FormulationGradientDivergenceTerms::SkewSymmetric)
+    {
+      do_evaluate_formulation<FormulationGradientDivergenceTerms::SkewSymmetric>(dst,
+                                                                                 src,
+                                                                                 time,
+                                                                                 zero_dst_vector);
+    }
+    else
+    {
+      AssertThrow(false, dealii::ExcMessage("Not implemented."));
+    }
```

Comment:
There might be more parameters than the weak vs. strong vs. skew-symmetric formulation that need to be evaluated in the innermost (q-point) loop. One example would be the type of flux (Lax-Friedrichs, Roe, central, ...). Since the number of branches in functions like this one would grow larger and larger, I am actually against the use of templates for such functionality.

Reply:
I think the length of the function should not be the reason to decide for or against the template parameter (this could even be generalized with `magic_enum::for_each` from the latest magic enum version as far as I understand).

I cannot reliably measure (I have only access to my laptop atm) if it has significant impact on the performance to remove the template parameter and use a standard `if` instead. I would at least avoid `if` in the innermost loop. In the case of the face loop, moving the `if` out of the loop over quadrature points leads to code duplication. On the other hand removing the template from the function would be better in terms of consistency, i.e. each existing operator could be more easily adjusted to the current design (also the ones derived from `OperatorBase`). 
For me this is a trade-off and since I am not able to do the measurements I will adjust to your wishes.

@nfehn You would prefer to remove the template from the function and have some code duplication, correct?

Reply:
What do you mean with "code duplication"? Currently, we have the same function call three times, apart from the template, which one could see as some form of code duplication. I think this could be removed by not using templates for parameters like `Formulation`.

Reply:
In essence, and since we do not have indications from performance measurements that these kind of if-statements are performance relevant, I would use templates only if necessary in the sense of generic programming and where inheritence / dynamic polymorphism is not an option (e.g. because `dealii::FEEValuation` would then have to derive from a base / interface class like `ExADG::ExteriorFaceIntegratorBase`, which is not possible). This holds for the `ExteriorFaceIntegrator` data types of this PR. This reasoning does not hold for the `Formulation` parameters, so I would not use templates for these parameters.

This is not meant as an absolute statement. It is meant as a pragmatic decision (as a compromise between performance  and code complexity) in the absence of performance numbers giving indications of potential bottlenecks.

Reply:
@nfehn What I meant with code duplication if we shift the if statement out of the innermost loop is the following:
```      
if (formulation == Formulation::Weak)
    for(unsigned int q : pressure_m.quadrature_point_indices())
    {
      scalar const pm = pressure_m.get_value(q);//duplicated
      scalar const pp = pressure_p.get_value(q);//duplicated
      vector const um = velocity_m.get_value(q);//duplicated
      vector const up = velocity_p.get_value(q);//duplicated
      vector const n  = pressure_m.normal_vector(q);//duplicated

      vector const flux_momentum =
        kernel.calculate_lax_friedrichs_flux_momentum(um, up, gamma, pm, pp, n);//duplicated

      scalar const flux_mass = kernel.calculate_lax_friedrichs_flux_mass(pm, pp, tau, um, up, n);//duplicated
      
      // formuation specific stuff
     }
else if (formulation == ...)
    for(unsigned int q : pressure_m.quadrature_point_indices())
    {
      scalar const pm = pressure_m.get_value(q);//duplicated
      scalar const pp = pressure_p.get_value(q);//duplicated
      vector const um = velocity_m.get_value(q);//duplicated
      vector const up = velocity_p.get_value(q);//duplicated
      vector const n  = pressure_m.normal_vector(q);//duplicated

      vector const flux_momentum =
        kernel.calculate_lax_friedrichs_flux_momentum(um, up, gamma, pm, pp, n);//duplicated

      scalar const flux_mass = kernel.calculate_lax_friedrichs_flux_mass(pm, pp, tau, um, up, n);//duplicated
      
      // formuation specific stuff
     }
```
From your response, I think you are suggesting the following, correct?
```
    for(unsigned int q : pressure_m.quadrature_point_indices())
    {
      scalar const pm = pressure_m.get_value(q);
      scalar const pp = pressure_p.get_value(q);
      vector const um = velocity_m.get_value(q);
      vector const up = velocity_p.get_value(q);
      vector const n  = pressure_m.normal_vector(q);

      vector const flux_momentum =
        kernel.calculate_lax_friedrichs_flux_momentum(um, up, gamma, pm, pp, n);

      scalar const flux_mass = kernel.calculate_lax_friedrichs_flux_mass(pm, pp, tau, um, up, n);

      if (formulation == Formulation::Weak)
          // formuation specific stuff
      else if (formulation == ...)
          // formuation specific stuff
```
I would like to propose the following: 
* We go with the second version from above
* I continue with a PR to be able to run a minimal example
* We test the performance for all three versions. If I have access to a cluster at that time, I am happy to do it, otherwise I might have to ask for help.
* If we can measure a noticeable performance difference, we reintroduce the template parameter

I made the changes as proposed. If you agree to above procedure, the PR is ready to merge from my side.

Reply:
Yes, this sounds like a good first step. We probably need to look into the variants a bit. I would personally not see templates as a form of code duplication (because it duplicates the machine code, not the maintained code), depending on how the code is setup up. The main part is that templates need a dispatch mechanism to select between the different variants at run-time, and how much work it is for the compiler to figure out different `if` statements compared to more compact statements.

As a side remark, I would point out that the compiler will have different heuristics that go in different directions, which can lead to "surprising" behavior where a template must not even uniformly lead to better performance, if the compiler e.g. decides to put common code into a function that comes with too much call overhead. Or that for dispatch is near the innermost part, like in the first variant you show here (but done via a templated inner function), the compile time or library size might not even be affected.

Reply:
We probably refer to different types of code duplication in this conversation. What I meant is the if-else statement with three times the same function call, the only difference being the template parameter.

@jh66637 I think we are of the same opinion regarding the face integrals (and that this is different than for the cell integrals). I was probably irritated because of the reference to conversation https://github.com/exadg/exadg/pull/587#discussion_r1377387760 (which discusses another aspect of "code duplication", which I did not mean in the present conversation).

Reply:
I think we are all on the same page :) 

I will add an issue to check the different versions and I think this PR can be merged, and I can continue adding a working acoustic solver :)